pkg/dockermanager/dockermanager.go

```diff
@@ -0,0 +1,197 @@
+	logging.Logging
+	ID      string
+	Manager *Manager
+}
+
+// Inspect inspects this docker container.
+func (dc *Container) Inspect(ctx context.Context) (types.ContainerJSON, error) {
+	return dc.Manager.ContainerInspect(ctx, dc.ID)
+}
+
+// IsOnline returns whether or not the container is online.
+func (dc *Container) IsOnline(ctx context.Context) (bool, error) {
+	info, err := dc.Inspect(ctx)
+	if err != nil {
+		if client.IsErrNotFound(err) {
+			err = nil
+		}
+		return false, err
+	}
+	switch info.ContainerJSONBase.State.Status {
+	case "running", "paused":
+		return true, nil
+	}
+	return false, nil
+}
+
+func (dc *Container) Exec(ctx context.Context, cmd ...string) error {
+	resp, err := dc.Manager.ContainerExecCreate(ctx, dc.ID, types.ExecConfig{
+		User:       "root",
+		Privileged: true,
+		Cmd:        cmd,
+	})
+	if err != nil {
+		return err
+	}
+	return dc.Manager.ContainerExecStart(ctx, resp.ID, types.ExecStartCheck{})
+}
+
+// Manage runs and stops workers as containers start and stop.
+//
+// If you pass labels, only containers labeled with at least one of the given
+// labels will be managed.
+func (dm *Manager) Manage(
+	ctx context.Context,
+	worker func(context.Context, *Container) error,
+	labels ...string,
+) error {
+
+	// Manage workers.
+	managers := make(map[string]struct {
+		done   chan struct{}
+		cancel context.CancelFunc
+	})
+
+	defer func() {
+		// cancel the remaining managers
+		for _, m := range managers {
+			m.cancel()
+		}
+		// wait for the running managers to exit
+		for _, m := range managers {
+			<-m.done
+		}
+	}()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	stop := func(container string) {
+		if m, ok := managers[container]; ok {
+			m.cancel()
+			delete(managers, container)
+			<-m.done
+		}
+	}
+	start := func(container string) {
+		if _, ok := managers[container]; ok {
+			return
+		}
+
+		cctx, cancel := context.WithCancel(ctx)
+		done := make(chan struct{})
+		managers[container] = struct {
+			done   chan struct{}
+			cancel context.CancelFunc
+		}{
+			done:   done,
+			cancel: cancel,
+		}
+		go func() {
+			defer close(done)
+			handle := dm.NewHandle(container)
+			err := worker(cctx, handle)
+			if err != nil {
+				handle.S().Errorf("sidecar worker failed: %s", err)
+			}
+		}()
+	}
+
+	// Manage existing containers.
+	now := time.Now()
+
+	listFilter := filters.NewArgs()
+	for _, l := range labels {
+		listFilter.Add("label", l)
+	}
+	nodes, err := dm.Client.ContainerList(ctx, types.ContainerListOptions{
+		Quiet:   true,
+		Limit:   -1,
+		Filters: listFilter,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	for _, n := range nodes {
+		start(n.ID)
+	}
+
+	eventFilter := listFilter.Clone()
+	eventFilter.Add("type", "container")
+	eventFilter.Add("event", "start")
+	eventFilter.Add("event", "stop")
+
+	// Manage new containers.
+	eventCh, errs := dm.Client.Events(ctx, types.EventsOptions{
+		Filters: eventFilter,
+		Since:   now.Format(time.RFC3339Nano),
```

Comment:
I was going to say this could end up managing the same container twice if it was created in the lapse of time between capturing `now` and acquiring the initial list, but I see that `start()` yields if the container is already managed. 👍